### ISSUE TITLE
Feature Request: Overriding Files from a Parent Chart

### ISSUE THREAD

**Issue by PengWang-SMARTM** [helm/helm#11293]
First of all, I'd believe this is a duplicated request of #1892 , but in a different wording.

Current Helm supports "Overriding Values from a Parent Chart", but apparently not Files yet.

As a result, one workaround is to convert the content of Files into Values, and make a huge values.yaml file to pass the configuration to subcharts. it works but is not very elegant.

A native support of "Overriding Files from a Parent Chart", if I understand correctly, would solve the issues mentioned in #1892 gracefully.

===
Some thoughts.

I'd guess, similar to the values, once the feature is implemented, files for subcharts should be organized in folders/directories named by the subcharts.

Since I'm new to helm/kubernetes, I'm still exploring the features and limits of helm/kubernetes. I feel I need a little more time and experiments to understand alias. I read some issues here, and helm seems having issues with supporting multi-instances of a same subchart but at different versions. I suspect, same or similar difficulties would be applicable to this feature as well.

**Comment by PengWang-SMARTM** [comment]
I believe #9420 is actually the same issue, rather than asking for accessing external files.

**Comment by joejulian** [comment]
I'm just not getting this. Could you perhaps reword this as a use-case?

**Comment by PengWang-SMARTM** [comment]
Let me try. It would be better if you can show me an example so I can follow a same style.

When [Subcharts ](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/) are used, the parent chart typically needs to override configuration for subcharts.

Such subchart configuration could be simple values as shown in the following example.
values.yaml in subchart:
```
image: "docker.example.com/subchart/image"
imageTag: "Tag1.0"
```

values.yaml in parent chart:
```
subchart:
  imageTag: "Tag2.0"
```

So the parent chart uses image `docker.example.com/subchart/image:Tag2.0` rather than `docker.example.com/subchart/image:Tag1.0`. Current helm supports this as "[Overriding Values from a Parent Chart](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/#overriding-values-from-a-parent-chart)"

However, subchart configuration could be saved in an external file in addition to `.Values`. For example:
configmap.yaml in subchart:
```
---
kind: ConfigMap
...
data:
  {{- (.Files.Glob .Values.dataFile).AsConfig | nindent 2 }}
```
values.yaml in subchart:
```
dataFile: "demoConfigDataFile"
```

Similar to above, I'd assume the values.yaml in parent chart should be something like:
```
subchart:
  dataFile: "productionConfigDataFile"
```
That is, the parent chart wants to use the configuration in `productionConfigDataFile` instead of `demoConfigDataFile`.

Since subchart doesn't have access to parent chart's `.Files`, we need helm to pass the `productionConfigDataFile `to the subchart. That's what I called "Overriding Files from a Parent Chart". To make things easier, it may be reasonable to add a restriction that only files under directory `subchart` can be passed to the subchart. That is, the directory would look like:
```
mychart/
  Chart.yaml
  values.yaml
  charts/
  templates/
  ...
  subchart/
    productionConfigDataFile
```



If so, someone may argue the example above for the parent chart needs to add the path to the file as shown below. To me, either way makes some sense, but the first one has less typing.
```
subchart:
  dataFile: "subchart\productionConfigDataFile"
```

**Comment by joejulian** [comment]
I'll un-stale this with a couple ideas.

What would happen if you created a file in `<parent-chart-root>/charts/subchart/Files`? Would that file end up in the subchart's Files?

What would happen if you inserted the file (base64 encoded) into `$.Subcharts.subchart.Files.<filename>`?

If someone with this problem has time to try it out, I'd be interested in hearing the results.

I solved this a different way. For the subchart, I set a flag `configmap.create.enabled` which enables or disables the creation of the ConfigMap. In my parent chart, I set that to false, `subchart.configmap.create.enabled=false` then in the parent chart, I template the subchart configmap and use data from the parent chart:

```
{{ include (print .Subcharts.subchart.Template.BasePath "/configmap.yaml") $myvalues }}
```

$myvalues is built to provide the values I want the template to have.

See https://github.com/redpanda-data/helm-charts/blob/main/charts/redpanda/templates/console/configmap.yaml for an example.

**Comment by PengWang-SMARTM** [comment]
> What would happen if you created a file in `<parent-chart-root>/charts/subchart/Files`? Would that file end up in the subchart's Files?

In our practice, we let "charts" directory be maintained by helm containing all dependencies (in .tgz files). The source of a subchart is either local files in <parent-chart-root>/<subchart>, or from a remote repository. When the subchart is local in <subchart>, the files there will be in both parent's .Files (unless in the .helmignore list) and subchart's .Files. If the subchart is from a remote repository, then files in <subchart> will be in only the parent's .Files.

I don't have the experiences with decompressed subcharts in the <parent-chart-root>/charts/ directory. However, according to [helm website](https://helm.sh/docs/chart_template_guide/accessing_files), .Files can access files in charts.

> 
> What would happen if you inserted the file (base64 encoded) into `$.Subcharts.subchart.Files.<filename>`?
By reading the accessing files web page above, I'd believe helm takes files as if they are regular txt files. If the file content is already base64 encoded, its content can be used directly when base64 is needed or need the b64dec to decode it.

> I solved this a different way. For the subchart, I set a flag `configmap.create.enabled` which enables or disables the creation of the ConfigMap. In my parent chart, I set that to false, `subchart.configmap.create.enabled=false` then in the parent chart, I template the subchart configmap and use data from the parent chart:
> 

Yes, this is pretty much I'm doing right now. There are some other workarounds. I personally like this one.

However, if the Overriding Files feature can be implemented, it would be even easier.

**Comment by joejulian** [comment]
If you wish to implement such a feature, you should probably start with a Helm Improvement Proposal ([HIP](https://github.com/helm/community/blob/main/hips/hip-0001.md)).

**Comment by PengWang-SMARTM** [comment]
> If you wish to implement such a feature, you should probably start with a Helm Improvement Proposal ([HIP](https://github.com/helm/community/blob/main/hips/hip-0001.md)).

Unfortunately, I'm steered away from Helm to a totally different area (again), which I'm not familiar with. I don't think I can find time in near future to work on this.

**Comment by joejulian** [comment]
Ok, I'll go ahead and close this. If circumstances change, feel free to reopen it.
